src/Storages/MergeTree/GinIndexStore.cpp

```diff
@@ -0,0 +1,435 @@
+
+        const auto& int_type = DB::DataTypePtr(std::make_shared<DB::DataTypeUInt32>());
+        auto size_serialization = int_type->getDefaultSerialization();
+        size_serialization->serializeBinary(result + n, *ostr);
+        ostr->sync();
+    }
+    return result;
+}
+
+UInt32 GinIndexStore::getNextRowIDRange(size_t n)
+{
+    UInt32 result =current_segment.next_row_id;
+    current_segment.next_row_id += n;
+    return result;
+}
+
+UInt32 GinIndexStore::getNextSegmentID()
+{
+    String sid_file_name = getName() + GIN_SEGMENT_ID_FILE_TYPE;
+    return getNextIDRange(sid_file_name, 1);
+}
+
+UInt32 GinIndexStore::getSegmentNum()
+{
+    if (cached_segment_num)
+        return cached_segment_num;
+
+    String sid_file_name = getName() + GIN_SEGMENT_ID_FILE_TYPE;
+    if (!storage->exists(sid_file_name))
+        return 0;
+    Int32 result = 0;
+    {
+        std::unique_ptr<DB::ReadBufferFromFileBase> istr = this->storage->readFile(sid_file_name, {}, std::nullopt, std::nullopt);
+
+        Field field_rows;
+        const auto& size_type = DB::DataTypePtr(std::make_shared<DB::DataTypeUInt32>());
+        auto size_serialization = size_type->getDefaultSerialization();
+
+        size_type->getDefaultSerialization()->deserializeBinary(field_rows, *istr);
+        result = field_rows.get<UInt32>();
+    }
+
+    cached_segment_num = result - 1;
+    return cached_segment_num;
+}
+
+bool GinIndexStore::needToWrite() const
+{
+    assert(max_digestion_size > 0);
+    return current_size > max_digestion_size;
+}
+
+void GinIndexStore::finalize()
+{
+    if (!current_postings.empty())
+    {
+        writeSegment();
+    }
+}
+
+void GinIndexStore::initFileStreams()
+{
+    String segment_file_name = getName() + GIN_SEGMENT_FILE_TYPE;
+    String item_dict_file_name = getName() + GIN_DICTIONARY_FILE_TYPE;
+    String postings_file_name = getName() + GIN_POSTINGS_FILE_TYPE;
+
+    segment_file_stream = data_part_storage_builder->writeFile(segment_file_name, DBMS_DEFAULT_BUFFER_SIZE, WriteMode::Append, {});
+    term_dict_file_stream = data_part_storage_builder->writeFile(item_dict_file_name, DBMS_DEFAULT_BUFFER_SIZE, WriteMode::Append, {});
+    postings_file_stream = data_part_storage_builder->writeFile(postings_file_name, DBMS_DEFAULT_BUFFER_SIZE, WriteMode::Append, {});
+}
+
+void GinIndexStore::writeSegment()
+{
+    if (segment_file_stream == nullptr)
+    {
+        initFileStreams();
+    }
+
+    ///write segment
+    segment_file_stream->write(reinterpret_cast<char*>(&current_segment), sizeof(GinIndexSegment));
+    std::vector<std::pair<std::string_view, GinIndexPostingsBuilderPtr>> token_postings_list_pairs;
+    token_postings_list_pairs.reserve(current_postings.size());
+
+    for (const auto& [token, postings_list] : current_postings)
+    {
+        token_postings_list_pairs.push_back({std::string_view(token), postings_list});
```

Comment:
Since the string which is referenced by string_view are in memory during the method, using string_view is an optimization by avoiding string copy.

Reply:
Right. What I meant was that the data structure we are doing `.push_back()` into is a vector of `std::pair<std::string_view, GinIndexPostingsBuilderPtr>`. So the conversion from `std::string` to `std::string_view` would happen automatically if we just write `token`. (this works because `string`s implicitly convert to `std::string_view`, https://en.cppreference.com/w/cpp/string/basic_string/operator_basic_string_view)